topdown/query.go

```diff
@@ -504,8 +524,21 @@ func (q *Query) Iter(ctx context.Context, iter func(QueryResult) error) error {
 		return iter(qr)
 	})
 
-	if q.strictBuiltinErrors && err == nil && len(e.builtinErrors.errs) > 0 {
-		err = e.builtinErrors.errs[0]
+	if len(e.builtinErrors.errs) > 0 {
+		if q.strictBuiltinErrors {
+			err = e.builtinErrors.errs[0]
+		} else if q.builtinErrorBuffer != nil {
+			for _, err := range e.builtinErrors.errs {
+				if tdError, ok := err.(*Error); ok {
+					*(q.builtinErrorBuffer) = append(*(q.builtinErrorBuffer), *tdError)
+				} else {
+					*(q.builtinErrorBuffer) = append(*(q.builtinErrorBuffer), Error{
+						Code:    BuiltinErr,
+						Message: err.Error(),
+					})
```

Comment:
[nit] This bit of logic got a little gnarly! :sweat_smile:  My only real worry was with this line (and its sister a few lines down):
```go
    *(q.builtinErrorBuffer) = append(*(q.builtinErrorBuffer), *tdError)
```
I wondered why the pointer-indirections were there, and then realized it was because of the pointer-to-slice type used earlier in the `Query` type's definition.

Does this logic (sans indirections) still work if you use just a plain `[]Error` slice type in the `Query` type, or will that change result in these mutations to the `builtinErrorBuffer` being lost? :thinking:

Reply:
>result in these mutations to the builtinErrorBuffer being lost

I think so, unless there's something I'm missing here... which is entirely possible!

Reply:
I think your reasoning is sound on why the slice pointer is there. Ugly happens sometimes!

I'd say for now, the best solution is probably just a quick comment documenting the _why_ behind the slice pointer indirections happening there. That way we won't have someone accidentally break it in the future with a well-meaning refactor, unless they're fixing the underlying problem.

Reply:
Thanks Philip, that makes sense. I have added some comments in d5d25d29186f279665d3be2ec915c85b2f75ce8f which I hope will make it clear what's going on here.